### PR TITLE
WIP: Spike a new Auth Policy for auth_clients

### PR DESCRIPTION
**Pull request by lyzadanger**
I wanted to share a partial approach for extending our authentication policies to bring `auth_client` authentication into the fold instead of handling it at the view level.

If I am on this branch, and have generated some `client_credentials` in my local `h` instance (for the `foobar.org` authority), I can get into the shell and run:

```
from h.auth import util
# Import my new policy that extends pyramid's BasicAuthAuthenticationPolicy
from h.auth.policy import AuthClientPolicy

# Base64-encoded CLIENT_ID, CLIENT_SECRET registered with my local `h`
request.headers['Authorization'] = "Basic MDQ0NjVhYWEtOGY3My0xMWU4LTkxY2EtOGJhMTE3NDJiMjQwOlcwaDN1cmtsbE00UjlYcE9GZjQyWUlhQ0Y2cEJLYkZTWDZ0SjcyR0tRYm8="
# `check` callback looks up auth_client in DB
basic = AuthClientPolicy(check=util.check_auth_client)
```

The above sets an `Authorization` header on the request (getting that to work probably represents 50% of the time invested on this!) and instantiates my new auth policy with a callback that reproduces the logic for verifying an auth_client against the database.

Now, I can:

```
>>> basic.effective_principals(request)
['system.Everyone', 'system.Authenticated', '04465aaa-8f73-11e8-91ca-8ba11742b240', 'auth_client:foobar.org', 'authority:foobar.org']
```

Note here that I've added a couple of principals (you can see this in the code), and that the "username" (CLIENT_ID) is in the principals, _and_ the request is considered "authenticated" (`system.Authenticated`), but, ta-da:

```
>>> basic.authenticated_userid(request)
>>>
```

No authenticated user! For now, you can do:

```
>>> basic.unauthenticated_userid(request)
>>> '04465aaa-8f73-11e8-91ca-8ba11742b240'
```

and you'll get the CLIENT_ID. For now I think this is OK.

I tracked things down, and our app won't set `request.user` if `request.authenticated_userid` isn't set. This is also ideal. (Took me a while to track down where this happens: `h.accounts.__init__`.

None of this is hooked up or tested at all, but I wanted to get eyes on it in case I'm doing something daft.

In summary: this auth policy gives us the ability to have an honest-to-god Pyramid Authenticated request, but one with out an `authenticated_userid` or a `user` attached to it. I'm hopeful this will give us the "hooks" we need to build useful things for view_configs to do the right kind of authorization.

Of course, there's work to be done still to integrate this into our existing auth policies and figure out the flow.

### CODE REVIEW

File: h/auth/util.py [hunk 1]
```diff
@@ -117,8 +117,63 @@ def authority(request):
     return text_type(request.registry.settings.get('h.authority', request.domain))
 
 
+def check_auth_client(username, password, request):
```
Comment (lyzadanger) [inline h/auth/util.py]: The actual logic within here is mostly a copy-and-paste job from the existing check.

File: h/auth/util.py [hunk 2]
```diff
@@ -117,8 +117,63 @@ def authority(request):
...
+
+    principals.add('auth_client:{authority}'.format(authority=client.authority))
+    principals.add('authority:{authority}'.format(authority=client.authority))
+
+    return list(principals)
+
+
 def request_auth_client(request):
     """
+    DEPRECATED
```
Comment (lyzadanger) [inline h/auth/util.py]: This was a note to self :)

**Comment by lyzadanger** [comment]
Note: I don't intend to get this particular branch to a mergeable state; just looking for feedback on the approach.

**Comment by seanh** [comment]
_Nice_, I really like the approach of having the authclient and `system.Authenticated` be in `effective_principals` but have `request.authenticated_userid` and `request.user` both be `None`. That neatly avoids a bunch of potential problems that we discussed.

Could there be any code anywhere that assumes that `system.Authenticated` being in `effective_principals` means that `request.authenticated_userid` and/or `request.user` are set? Would this be a safe assumption for code to make anyway, under normal (non-authclient) circumstances?

Is a little bit weird that `unauthenticated_userid` is set but `authenticated_userid` isn't? I think that might actually be perfectly sensible. For example it's a perfectly normal Pyramid thing for a request to come in with a valid signed cookie or JWT or auth header or what have you declaring a certain userid, causing `unauthenticated_userid` to be set to that userid, but then that user doesn't actually exist in the DB (e.g. the user account has been deleted) so `authenticated_userid` ends up `None`. This isn't exactly the same situation but it seems sort of analogous,

So the `'04465aaa-8f73-11e8-91ca-8ba11742b240'` in `effective_principals` is the `AuthClient.id` (a.k.a. the OAuth 2.0 `client_id`)? Is it necessary to have that in `effective_principals` as well as `'auth_client:foobar.org'`? Related, I was wondering whether ` 'auth_client:foobar.org'` is unique or can you have two different `foobar.org` auth clients? (May not be important.)

We're doing something slightly off the garden path here with auth clients and this seems like a good approach to getting it to work.

**Comment by lyzadanger** [comment]
Thanks for the feedback! I'm going to close this PR now and continue work on implementing this approach.
